src/data/apint.rs

```diff
@@ -0,0 +1,231 @@
+            // (At the end of this method.)
+            self.as_digit_slice_mut()
+                .copy_from_slice(rhs.as_digit_slice());
+        }
+        else {
+            // In this case `rhs` and `self` require an unequal amount
+            // of digits for their representation which means that the
+            // digits that may be allocated by `self` must be dropped.
+            //
+            // Note that `ApInt::drop_digits` only deallocates if possible.
+            unsafe{ self.drop_digits(); }
+
+            match rhs.storage() {
+                Storage::Inl => {
+                    // If `rhs` is a small `ApInt` we can simply update
+                    // the `digit` field of `self` and we are done.
+                    self.data.inl = unsafe{ rhs.data.inl };
+                }
+                Storage::Ext => {
+                    // If `rhs` is a large heap-allocated `ApInt` we first
+                    // need to expensively clone its buffer and feed it to `self`.
+                    let cloned = rhs.clone();
+                    self.data.ext = unsafe{ cloned.data.ext };
+                    use std::mem;
+                    mem::forget(cloned);
+                }
+            }
+        }
+        // Since all cases may need bit width adjustment we outsourced it
+        // to the end of this method.
+        self.len = rhs.len;
+    }
+
+    /// Strictly assigns `rhs` to this `ApInt`.
+    /// 
+    /// After this operation `rhs` and `self` are equal to each other.
+    /// 
+    /// **Note:** Strict assigns protect against mutating the bit width
+    /// of `self` and thus return an error instead of executing a probably
+    /// expensive `assign` operation.
+    /// 
+    /// # Errors
+    /// 
+    /// - If `rhs` and `self` have unmatching bit widths.
+    pub fn strict_assign(&mut self, rhs: &ApInt) -> Result<()> {
+        if self.width() != rhs.width() {
+            return Error::unmatching_bitwidths(self.width(), rhs.width())
+                .with_annotation(format!(
+                    "Occured while trying to `strict_assign` {:?} to {:?}.", self, rhs))
+                .into()
+        }
+        self.as_digit_slice_mut()
+            .copy_from_slice(rhs.as_digit_slice());
+        Ok(())
+    }
+}
+
+impl Drop for ApInt {
+    fn drop(&mut self) {
+        unsafe{self.drop_digits()}
+    }
+}
+
+//we think that this compiles on `-O3` down to the optimal code
+impl Clone for ApInt {
+    fn clone(&self) -> Self {
+        match self.storage() {
+            Storage::Inl => {
+                ApInt::new_inl(self.len, unsafe{ self.data.inl })
+            }
+            Storage::Ext => {
+                use std::mem;
+                let req_digits = self.len_digits();
+                let mut buffer = self.as_digit_slice()
+                    .to_vec()
+                    .into_boxed_slice();
+                assert_eq!(buffer.len(), req_digits);
```

Comment:
I think with https://doc.rust-lang.org/std/vec/struct.Vec.html#method.reserve_exact we can do the same.
I am not sure if it is guaranteed that our current approach internally calls https://doc.rust-lang.org/std/vec/struct.Vec.html#method.reserve_exact.